### PR TITLE
Added SiLU activation function

### LINKED ISSUE #3169: [Feature Request] Swish Activation Function
Swish ([arxiv](https://arxiv.org/pdf/1710.05941.pdf)) is an activation function that has been shown to empirically outperform ReLU and several other popular activation functions on Inception-ResNet-v2 and MobileNet. On models with more layers Swish typically outperforms ReLU. Implementation is simple:

swish

Sigma is just sigmoid.

Worth a PR?

cc @albanD @mruberry

### PR DESCRIPTION
**Pull request by heitorschueroff**
Implemented the SiLU activation function as discussed in #3169.

### CODE REVIEW

File: aten/src/ATen/native/Activation.cpp [hunk 1]
```diff
@@ -172,6 +172,19 @@ Tensor & relu_(Tensor & self) {
...
+
+Tensor & swish_(Tensor& self) {
+  return self.mul_(self.sigmoid_());
+}
+
+Tensor swish_backward(const Tensor& grad, const Tensor& self) {
+  auto self_sigmoid = self.sigmoid();
+  return grad * (self_sigmoid * (1 + self * (1 - self_sigmoid)));
+}
+
```
Comment (glaringlee) [inline aten/src/ATen/native/Activation.cpp:187]: Please add a .out version, here is an example
https://github.com/pytorch/pytorch/pull/41034/files#diff-8026a1a389bfcf92e484ae272675c336R1187

File: tools/autograd/derivatives.yaml [hunk 2]
```diff
@@ -1181,6 +1181,12 @@
 - name: relu_(Tensor(a!) self) -> Tensor(a!)
   self: threshold_backward(grad, result, 0)
 
+- name: swish(Tensor self) -> Tensor
+  self: swish_backward(grad, self)
+
+- name: swish_(Tensor(a!) self) -> Tensor(a!)
```
Comment (glaringlee) [inline tools/autograd/derivatives.yaml:1187]: @albanD 
Will this work? For inplece, since the self here is not original input, but forward of input, it seems to me that this won't work.

**Review by glaringlee — CHANGES_REQUESTED** [review]
@albanD 
This is an initial draft, should this be put in nn?

**Comment by vadimkantorov** [comment]
Will there be a `torch.nn.Swish` module as well? (hopefully with `inplace = True` support)

**Comment by albanD** [comment]
I think that if you want to have the full `nn` experience, we should add it to `nn.functional` and a Module that uses the functional version yes.
But if you want to leave it as a follow up up to keep this one small as it is a first task, it is fine as well!

**Comment by TFUsers** [comment]
Please modify this pull request to rename the swish to the SiLU.

The swish was originally coined the "SiLU" in https://arxiv.org/pdf/1606.08415.pdf and https://arxiv.org/abs/1702.03118 long before the swish paper. Renaming other peoples' exact same ideas is unacceptable.
This request inspired by a [recent discussion](https://www.reddit.com/r/MachineLearning/comments/hkiyir/r_google_has_a_credit_assignment_problem_in/) and a recent [jax](https://github.com/google/jax/issues/3672) and [tensorflow issue](https://github.com/tensorflow/tensorflow/issues/41066), but this problem has been brought up every few months for the past few years. In light of recent efforts to make the ML community more equitable and _fair_, this is a no-brainer and long overdue.
PyTorch hasn't added the swish precisely because its renaming was unfair (just search previous pull requests). Changing the swish to the SiLU, as Jax just did and tensorflow will do, will put this problem to rest.

File: tools/autograd/derivatives.yaml (hunk 2 above)
Comment (albanD) [reply tools/autograd/derivatives.yaml:1187]: After discussion, this is not possible as we can't compute the derivative without the original value :'(  (the function is not bijective) cc @vadimkantorov 
So you can replace it with `not_implemented("swish_")`.
Not sure how to add a note here to mention that it is not possible to do so, maybe `not_implemented("swish_ (inplace version cannot compute the gradient, use out of place)")` will give a better error message.

**Comment by albanD** [comment]
That sounds like a good point @TFUsers 

@heitorschueroff could you make the following changes then:
- Change all the namings in the current PR from `swish` to `silu`.
- Add in `torch/functional.py` a new function like:
```
def swish(inp):
  """
  Some doc explaining it is called silu here and in torch.nn
  """
  return torch.silu(inp)
```
This is just to make sure that people looking for swish in our documentation page hit something that points them to what they're looking for.

Also the doc is missing from your PR:
- Add entries for `silu` and `silu_` in `torch/_tensor_docs.py` similar to the sigmoid ones.
- Add entries for `silu` in `torch/_torch_docs.py` similar to the sigmoid one.
- Make sure to add entries in `docs/source/torch.rst` next to sigmoid for both `swish` and `silu`.
- Make sure to add entries in `docs/source/tensor.rst` for `silu` and `silu_`.

File: aten/src/ATen/native/Activation.cpp [hunk 3]
```diff
@@ -172,6 +172,19 @@ Tensor & relu_(Tensor & self) {
...
 
+Tensor swish(const Tensor& self) {
+  return self.mul(self.sigmoid());
+}
+
+Tensor & swish_(Tensor& self) {
+  return self.mul_(self.sigmoid_());
+}
+
+Tensor swish_backward(const Tensor& grad, const Tensor& self) {
```
Comment (albanD) [inline aten/src/ATen/native/Activation.cpp:183]: Note that you don't need to make this an actual function in native_functions (that gets exposed into the `torch.*` namespace and that you need to document, etc)
Usually, I would add this function to [this file](https://github.com/pytorch/pytorch/blob/a78024476ba030b3439726eb96ab55a23a38a453/tools/autograd/templates/Functions.cpp) (remove the entry in native_function.yaml as well).
And you can leave the derivatives.yaml definition as-is and it will work just fine.

**Comment by gchanan** [comment]
CC @mruberry: we have an alias mechanism, right?  That we can use to alias swish to silu.

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -2336,6 +2336,16 @@
     MkldnnCPU: mkldnn_relu_
     QuantizedCPU: quantized_relu_
 
+- func: swish(Tensor self) -> Tensor
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml:2339]: should swish have `python_module: nn` ?

File: aten/src/ATen/native/Activation.cpp (hunk 3 above)
Comment (gchanan) [reply aten/src/ATen/native/Activation.cpp:183]: fwiw functions with _backward don't get bound automatically.

**Comment by mruberry** [comment]
> CC @mruberry: we have an alias mechanism, right? That we can use to alias swish to silu.

Not yet. File a follow-up issue and assign it to me and I'll add it when we finalize aliasing in a few weeks.

**Comment by gchanan** [comment]
this is related to `pytorch_module: nn`, but I don't think we need to make this a method; relu isn't a method, for example.

**Comment by glaringlee** [comment]
@gchanan 
relu has torch.relu, check here
https://github.com/pytorch/pytorch/blob/5e03a1e9261e86d3ee62481ef40d98ca5f09646c/torch/nn/functional.py#L294
Should we add python_module:nn and add silu into nn/functional.py?
otherwise, silu can only be accessed through torch._C._nn.silu
or
Should we remove python_module:nn and just provide torch.silu?

**Comment by hendrycks** [comment]
> Should we remove python_module:nn and just provide torch.silu?
Originally torch.gelu (the activation function used in BERT) was provided, but then it became F.gelu.

**Comment by glaringlee** [comment]
To be consistent,  to be consistent, let's do nn.functional.silu (F.silu)

**Review by glaringlee — CHANGES_REQUESTED** [review]
So far looks good to me.
There are several things that need to add before ship this:
1. For documents, please update the following two rst file to make it indexed.
   a. docs/source/nn.functional.rst
       In 'Non-linear activation functions' section, probably after `hardsigmoid`
   b. docs/source/nn.rst
       In `Non-linear Activations (weighted sum, nonlinearity) section, add SiLU in
   here is what it should looks like (you can click nn.SiLU and nn.functional.silu)
   https://glaringlee.github.io/search.html?q=silu&check_keywords=yes&area=default

   After the rst file get updated, you can try build the doc locally. Here is how.
   https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md#writing-documentation
2. Update torch/_override.py
    Add silu into the list of get_testing_overrides(), this is the root cause of the CI test failure.

File: torch/_overrides.py [hunk 5]
```diff
@@ -612,6 +612,7 @@ def get_testing_overrides():
         torch.selu: lambda input, inplace=False: -1,
         torch.sigmoid: lambda input, out=None: -1,
         torch.sign: lambda input, out=None: -1,
+        torch.silu: lambda input, out=None: -1,
```
Comment (glaringlee) [inline torch/_overrides.py:615]: this should be torch.nn.functional.silu since silu is under nn functional.

**Review by glaringlee — CHANGES_REQUESTED** [review]

**Review by glaringlee — APPROVED** [review]
LGTM now.

**Comment by iiSeymour** [comment]
Did this not make it into 1.6 in the end?

**Comment by albanD** [comment]
The original branch cut happened on the 23rd of July IIRC. And then only important bugfix are cherry picked into the release.
So I don't think so :/
